### PR TITLE
feat(microservices): allow kafka serializer to be async and await result

### PR DESCRIPTION
**Pull request by guiruiz**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Microservices options allow configuration of custom serializers but currently it doesn't support async serialize functions.

Issue Number: N/A

## What is the new behavior?
From now serializer functions can be async and if so the serialization result will be awaited

## Does this PR introduce a breaking change?
- [ ] Yes
- [X] No

## Other information
I had the need to serialize messages using Avro Schemas with the Schema Registry that relies on async functions.
In order to make it work on my project besides creating a custom serializer I had to extend ServerKafka and ClientKafka class to support async serialize functions.
Since this is an issue that other people may have I decided to contribute opening this PR.
This change does not introduce a breaking change because calling `await syncFunction()` will simply return the function result.

### CODE REVIEW

**Review by JoshVee — APPROVED** [review]

File: packages/microservices/client/client-kafka.ts [hunk 1]
```diff
@@ -222,23 +223,27 @@ export class ClientKafka extends ClientProxy {
...
-      this.producer.send(message).catch(err => callback({ err }));
+      this.serializer
+        .serialize(packet.data)
+        .then((serializedPacket: KafkaRequest) => {
+          serializedPacket.headers[KafkaHeaders.CORRELATION_ID] = packet.id;
+          serializedPacket.headers[KafkaHeaders.REPLY_TOPIC] = replyTopic;
+          serializedPacket.headers[KafkaHeaders.REPLY_PARTITION] =
+            replyPartition;
+
+          this.routingMap.set(packet.id, callback);
```
Comment (kamilmysliwiec) [inline packages/microservices/client/client-kafka.ts:234]: If someone unsubscribes before this promise resolves, a memory leak would occur (`this.routingMap.set(packet.id, callback);`)

File: packages/microservices/client/client-kafka.ts [hunk 2]
```diff
@@ -189,17 +189,18 @@ export class ClientKafka extends ClientProxy {
...
     const message = Object.assign(
       {
         topic: pattern,
         messages: [outgoingEvent],
       },
       this.options.send || {},
     );
-    return this.producer.send(message);
+
+    return await this.producer.send(message);
```
Comment (kamilmysliwiec) [inline packages/microservices/client/client-kafka.ts:203]:
```suggestion
    return this.producer.send(message);
```
You don't need `await` here

File: packages/microservices/server/server-kafka.ts [hunk 3]
```diff
@@ -140,8 +140,8 @@ export class ServerKafka extends Server implements CustomTransportStrategy {
     replyPartition: string,
     correlationId: string,
   ): (data: any) => Promise<RecordMetadata[]> {
-    return (data: any) =>
-      this.sendMessage(data, replyTopic, replyPartition, correlationId);
+    return async (data: any) =>
+      await this.sendMessage(data, replyTopic, replyPartition, correlationId);
```
Comment (kamilmysliwiec) [inline packages/microservices/server/server-kafka.ts:144]:
```suggestion
     this.sendMessage(data, replyTopic, replyPartition, correlationId);
```

You don't need `await` here

File: packages/microservices/server/server-kafka.ts [hunk 4]
```diff
@@ -207,7 +207,7 @@ export class ServerKafka extends Server implements CustomTransportStrategy {
       },
       this.options.send || {},
     );
-    return this.producer.send(replyMessage);
+    return await this.producer.send(replyMessage);
```
Comment (kamilmysliwiec) [inline packages/microservices/server/server-kafka.ts:210]:
```suggestion
    return this.producer.send(replyMessage);
```

You don't need `await` here

**Review by JoshVee — APPROVED** [review]

File: packages/microservices/client/client-kafka.ts (hunk 1 above)
Comment (guiruiz) [reply packages/microservices/client/client-kafka.ts]: Yeah, you were right.
So I implemented a flag that disables the routingMap inclusion if someone unsubscribes before the promise resolves.

Don't you think that a memory leak could occur with the current prod code (before these code changes)?

Comment (kamilmysliwiec) [reply packages/microservices/client/client-kafka.ts]: @guiruiz it wouldn't because `serialize` isn't async and so `this.routingMap.set` is called synchronously.

Comment (guiruiz) [reply packages/microservices/client/client-kafka.ts]: Hmm, that makes sense, but there is a chance to the route be included in the routingMap and the observer.complete never be called, isn't there? 

BTW, what do you think about my solution to avoid the memory leak in an async approach?

Comment (guiruiz) [reply packages/microservices/client/client-kafka.ts]: @kamilmysliwiec I simplified a bit the solution and wrote a unit test for this scenario.
Let me know your thoughts about it.

Comment (guiruiz) [reply packages/microservices/client/client-kafka.ts]: Hey @kamilmysliwiec, any thoughts about it?

File: packages/microservices/serializers/kafka-request.serializer.ts [hunk 5]
```diff
@@ -13,8 +13,10 @@ export interface KafkaRequest<T = any> {
   headers: Record<string, any>;
 }
 
-export class KafkaRequestSerializer implements Serializer<any, KafkaRequest> {
-  serialize(value: any): KafkaRequest {
+export class KafkaRequestSerializer
+  implements Serializer<any, KafkaRequest | Promise<KafkaRequest>>
+{
+  async serialize(value: any): Promise<KafkaRequest> {
```
Comment (micalevisk) [inline packages/microservices/serializers/kafka-request.serializer.ts:19]: what do you think on adding a JSDoc comment here to clarify the usage of `async` keyword (or the return of a promise) as there's no `await` inside it?

File: packages/microservices/client/client-kafka.ts [hunk 6]
```diff
@@ -216,33 +217,47 @@ export class ClientKafka extends ClientProxy {
...
-          messages: [serializedPacket],
-        },
-        this.options.send || {},
-      );
-      this.producer.send(message).catch(err => callback({ err }));
-
-      return () => this.routingMap.delete(packet.id);
+      this.serializer
+        .serialize(packet.data)
+        .then((serializedPacket: KafkaRequest) => {
```
Comment (cikay) [inline packages/microservices/client/client-kafka.ts:240]: Use async-await feature rather than then and catch

Comment (guiruiz) [reply packages/microservices/client/client-kafka.ts:240]: In order to use async-await feature here we'd have to make `publish` function async and modify all of its stack calls to use async-await.

File: packages/microservices/serializers/kafka-request.serializer.ts (hunk 5 above)
Comment (guiruiz) [reply packages/microservices/serializers/kafka-request.serializer.ts]: Yeah you're right, there's no need for the serialize function be async.

So I removed it from serialize function but kept the `Promise<KafkaRequest>` on the `Serializer` interface in case of someone wants to extend the class KafkaRequestSerializer overriding the serialize function.

Comment (micalevisk) [reply packages/microservices/serializers/kafka-request.serializer.ts]: wouldn't `KafkaRequest | Promise<KafkaRequest>` be more aligned to the implemented interface

https://github.com/nestjs/nest/blob/28e37696257a3f3fca1d95b5e692933370ed1d0c/packages/microservices/interfaces/serializer.interface.ts#L7-L9

Comment (guiruiz) [reply packages/microservices/serializers/kafka-request.serializer.ts]: Dropping the return type makes more sense to me. It's done 😄

Comment (guiruiz) [reply packages/microservices/serializers/kafka-request.serializer.ts]: The test pipeline is passing now 🤙

**Review by xSerafim — APPROVED** [review]

**Comment by guiruiz** [comment]
@kamilmysliwiec @micalevisk Hey guys, any other considerations to merge this PR?

File: packages/microservices/client/client-kafka.ts [hunk 7]
```diff
@@ -1,6 +1,7 @@
 import { Logger } from '@nestjs/common/services/logger.service';
 import { loadPackage } from '@nestjs/common/utils/load-package.util';
 import { isUndefined } from '@nestjs/common/utils/shared.utils';
+import { resolve } from 'path/posix';
```
Comment (kamilmysliwiec) [inline packages/microservices/client/client-kafka.ts:4]: Why is this needed?

Comment (guiruiz) [reply packages/microservices/client/client-kafka.ts]: @kamilmysliwiec  Actually it's not needed, I just removed it.

**Comment by kamilmysliwiec** [comment]
lgtm
